[#60562] Consolidate "Download PDF" with "Generate PDF"

as-op opened the pull request:
# Ticket

https://community.openproject.org/work_packages/60562

# What are you trying to achieve?

Consolidate "Download PDF" feature with the "Generate PDF" feature in one export modal and add an admin section for work package types to enable/disable/sort PDF export templates

## Screenshots

### Modal
<img width="535" alt="Bildschirmfoto 2025-02-18 um 15 05 30" src="https://github.com/user-attachments/assets/f8fdac4f-d13b-40e9-9e61-e850d06af689" />

### Settings
<img width="921" alt="Bildschirmfoto 2025-02-18 um 15 04 38" src="https://github.com/user-attachments/assets/3ea2fecf-36ab-4141-996d-eca261b6a207" />

# What approach did you choose and why?

* Use turbo & primer for the modal
* Use JSONB for the storage of the pdf export template settings for a type
   * a list of disabled templates (so no need for a migration to enable existing and future new ones)
   * a sorted list (only if the order has been modified at least once)

# Merge checklist

- [x] Remove the "Download PDF" menu entry
- [x] Add template selection to the export modal dialog
- [x] template selection displays a caption depending on the selected 
- [x] Split hyphenation "Off" from selection into a checkbox
- [x] Auto-select hyphenation language from the current language if available
- [x] Remove paper size select   
- [x] Apply footer text settings to work package exporter
- [x] Remove the feature flag 
- [x] Add template selection to work package types admin section
   - [x] Add data to Type
   - [x] Allow manual sorting of templates in type
   - [x] Allow deactivating templates in type   
- [x] Filter and order template selection in the dialog by work package type
- [x] Added/updated tests
- [x] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

as-op committed:
[#60562] Consolidate "Download PDF" feature with the "Generate PDF" feature in one export modal

https://community.openproject.org/work_packages/60562

as-op committed:
obey linters

as-op committed:
obey erb_lint

as-op committed:
* change caption for changed template
* remove feature flag
* adjust dialog spec

as-op committed:
obey rubocop

as-op committed:
restore pdf link for API

as-op committed:
adjust spec to match changed parameters

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
add admin settings tab for pdf export templates

as-op committed:
obey rubocop

as-op committed:
add missing change

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
remove header custom text, switch footer text inputs by template selection

as-op committed:
obey rubocop

as-op committed:
fix specs for export footer changes

as-op committed:
fix specs for export footer changes

as-op committed:
obey rubocop

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
fix specs

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
connect the settings UI with the model

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
refactor spread code into a type sub class

as-op committed:
obey rubocop & erb_lint

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
fix missing refactoring

as-op committed:
Merge branch 'dev' into feature/60562-consolidate-download-pdf-feature-with-the-generate-pdf-feature-in-one-export-modal

as-op committed:
add controller spec

as-op committed:
add frontend spec

klaustopher commented on app/models/type/pdf_export_templates.rb:
```diff
@@ -0,0 +1,93 @@
+  end
+
+  def find(template_id)
+    build_in_template = build_in_templates.find { |t| t[:id] == template_id }
```
Just a naming suggestion: You mean the templates that are built in? Then it should be `built_in_templates` otherwise it would be the templates that we do a build in :D

klaustopher reviewed (approved):
Added one remark about naming. Tried it out in the browser, works as intended, tried to break it by adding weird stuff to the database. All good 👍

as-op committed:
naming fix: build_in => built_in
resolves https://github.com/opf/openproject/pull/17840#pullrequestreview-2626598219
